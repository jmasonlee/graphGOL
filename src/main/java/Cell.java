public class Cell implements Comparable {
    public Integer y;
    public Integer x;

    public Cell(Integer x, Integer y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(Object o) {
        return 0;
    }
}
