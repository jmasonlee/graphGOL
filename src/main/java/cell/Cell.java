package cell;

public class Cell implements Comparable {
  public Integer y;
  public Integer x;

  public Cell(Integer x, Integer y) {
    this.x = x;
    this.y = y;
  }

  @Override
  public int compareTo(Object o) {
    return 0;
  }

  @Override
  public boolean equals(Object o) {

    if (o == this) {
      return true;
    }

    if (!(o instanceof Cell)) {
      return false;
    }

    Cell otherCell = (Cell) o;

    return otherCell.x == this.x && otherCell.y == this.y;
  }

  @Override
  public String toString(){
      return "[" + x + "," + y + "]";
  }
}
