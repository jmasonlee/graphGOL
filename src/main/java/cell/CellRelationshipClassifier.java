package cell;

public class CellRelationshipClassifier {
    public static Relationships classify(Cell cell, Cell cell1) {
        return Relationships.SELF;
    }

    public enum Relationships {
        DISCONNECTED(-1), SELF(0), NEIGHBOUR(1), COPARENT(2);

        public final int maxDistance;

        private Relationships(int maxDistance){
            this.maxDistance = maxDistance;
        }
    }


}
