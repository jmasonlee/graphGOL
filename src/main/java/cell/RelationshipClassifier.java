package cell;

public class RelationshipClassifier {
    public static Relationships classify(Cell cell, Cell cell1) {
        return Relationships.SELF;
    }

    public enum Relationships {
        DISCONNECTED(-1), SELF(0), NEIGHBOUR(1), COPARENT(2);

        public final int maxDistance;

        private Relationships(int maxDistance){
            this.maxDistance = maxDistance;
        }
    }


}
