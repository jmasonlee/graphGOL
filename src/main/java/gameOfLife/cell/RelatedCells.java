package gameOfLife.cell;

import java.util.List;
import java.util.Map;

public class RelatedCells {
  private final Map<Relationships, List<Cell>> relationships;

  public RelatedCells(Map<Relationships, List<Cell>> relationships) {
    this.relationships = relationships;
  }

  public Map<Relationships, List<Cell>> getRelationships() {
    return relationships;
  }

  @Override
  public String toString() {
    return relationships.toString();
  }

  public List<Cell> getCellsOfRelationshipType(Relationships relationshipType) {
    return relationships.get(relationshipType);
  }
}
