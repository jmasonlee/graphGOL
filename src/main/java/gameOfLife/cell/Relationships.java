package gameOfLife.cell;

public enum Relationships {
  DISCONNECTED(-1),
  SELF(0),
  NEIGHBOUR(Math.sqrt(2)),
  COPARENT(Math.sqrt(8));

  public final double maxDistance;

  Relationships(double maxDistance) {
    this.maxDistance = maxDistance;
  }
}
