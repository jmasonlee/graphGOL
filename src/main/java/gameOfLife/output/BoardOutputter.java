package gameOfLife.output;

import gameOfLife.cell.Cell;
import org.apache.commons.lang.StringUtils;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

public class BoardOutputter {

  private static final String TOP_BOUNDARY = " __";

  public static String createBoardOutput(List<Cell> cells) {
    BoardBounds boardBounds = new BoardBounds(cells);
    StringBuilder emptyBoard = emptyBoardOfSize(boardBounds.getWidth(), boardBounds.getHeight());
    return populateBoardWithCells(emptyBoard, cells, boardBounds);
  }

  private static String populateBoardWithCells(StringBuilder board, List<Cell> cells, BoardBounds boardBounds) {
    cells.forEach(c -> {
      int positionInBoard = getCellPositionInBoard(c, boardBounds);
      board.setCharAt(positionInBoard, 'X');
    });
    return board.toString();
  }

  private static int getCellPositionInBoard(Cell cell, BoardBounds boardBounds) {
    int adjustedX = Math.abs(boardBounds.getLeftmostXValue() - cell.x);
    int adjustedY = Math.abs(boardBounds.getUpperYValue() - cell.y);

    int boardWidth = (3 * boardBounds.getWidth()) + 2;

    return boardWidth + (adjustedY * boardWidth) + (adjustedX * 3);
  }

  private static StringBuilder emptyBoardOfSize(int width, int height) {
    StringBuilder board = new StringBuilder();

    board.append(StringUtils.repeat(TOP_BOUNDARY, width));
    board.append("\n");

    for (int i = 0; i < height; i++) {
      board.append(StringUtils.repeat("|__", width));
      board.append("|\n");
    }

    return board;
  }

  public static class BoardBounds {
    public static final int DEFAULT_BOARD_SIZE = 5;
    private int upperYValue = 5;
    private int leftmostXValue = 0;
    private int height = DEFAULT_BOARD_SIZE;
    private int width = DEFAULT_BOARD_SIZE;

    public BoardBounds(List<Cell> cellsForBoard) {
      calculateBoardBounds(cellsForBoard);
    }

    public int getWidth() {
      return width;
    }

    public int getHeight() {
      return height;
    }

    public int getUpperYValue() {
      return upperYValue;
    }

    public int getLeftmostXValue() {
      return leftmostXValue;
    }

    private void calculateBoardBounds(List<Cell> cellsForBoard) {
      if (cellsForBoard.isEmpty()) {
        return;
      }

      Function<Cell, Integer> fetchYvalue = c->c.y;
      upperYValue = getCellWithMaxDimension(cellsForBoard, fetchYvalue).y;
      int lowerYValue = getCellWithMinDimension(cellsForBoard, fetchYvalue).y;
      height = getDimension(lowerYValue, upperYValue);

      Function<Cell, Integer> fetchXvalue = c->c.x;
      leftmostXValue = getCellWithMinDimension(cellsForBoard, fetchXvalue).x;
      int rightmostXValue = getCellWithMaxDimension(cellsForBoard, fetchXvalue).x;
      width = getDimension(rightmostXValue, leftmostXValue);
    }

    private Cell getCellWithMinDimension(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCellDimension) {
      return Collections.min(cellsForBoard, Comparator.comparing(c -> fetchCellDimension.apply(c)));
    }

    private Cell getCellWithMaxDimension(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCellDimension) {
      return Collections.max(cellsForBoard, Comparator.comparing(c->fetchCellDimension.apply(c)));
    }

    private int getDimension(int lowestDimensionValue, int highestDimensionValue) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) < DEFAULT_BOARD_SIZE ?
          DEFAULT_BOARD_SIZE :
          calculateDimensionLargerThanDefault(lowestDimensionValue, highestDimensionValue);
    }

    private int calculateDimensionLargerThanDefault(int lowestDimensionValue, int highestDimensionValue) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) + 1;
    }

  }
}