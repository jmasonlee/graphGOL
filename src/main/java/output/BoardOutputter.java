package output;

import cell.Cell;
import org.apache.commons.lang.StringUtils;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class BoardOutputter {

  public static String createBoardOutput(List<Cell> cells) {
    BoardBounds boardBounds = new BoardBounds(cells);
    StringBuilder emptyBoard = emptyBoardOfSize(boardBounds.getWidth(), boardBounds.getHeight());
    return populateBoardWithCells(emptyBoard, cells, boardBounds);
  }

  private static String populateBoardWithCells(StringBuilder board, List<Cell> cells, BoardBounds boardBounds) {
    cells.forEach(c -> {
      int positionInBoard = getCellPositionInBoard(c, boardBounds);
      board.setCharAt(positionInBoard, 'X');
    });
    return board.toString();
  }

  private static int getCellPositionInBoard(Cell cell, BoardBounds boardBounds) {
    int adjustedX = Math.abs(boardBounds.getLeftmostXValue() - cell.x);
    int adjustedY = Math.abs(boardBounds.getUpperYValue() - cell.y);
    return (adjustedY * boardBounds.getWidth()) + adjustedY + adjustedX;
  }

  private static StringBuilder emptyBoardOfSize(int width, int height) {
    StringBuilder board = new StringBuilder();

    for (int i = 0; i < height; i++) {
      board.append(StringUtils.repeat(" ", width));
      if (i < height - 1) {
        board.append("\n");
      }
    }

    return board;
  }

  public static class BoardBounds {
    public static final int DEFAULT_BOARD_SIZE = 5;
    private int upperYValue = 5;
    private int leftmostXValue = 0;
    private int height = DEFAULT_BOARD_SIZE;
    private int width = DEFAULT_BOARD_SIZE;

    public BoardBounds(List<Cell> cellsForBoard) {
      calculateBoardBounds(cellsForBoard);
    }

    public int getWidth() {
      return width;
    }

    public int getHeight() {
      return height;
    }

    public int getUpperYValue() {
      return upperYValue;
    }

    public int getLeftmostXValue() {
      return leftmostXValue;
    }

    private void calculateBoardBounds(List<Cell> cellsForBoard) {
      if (cellsForBoard.isEmpty()) {
        return;
      }

      upperYValue = Collections.max(cellsForBoard, Comparator.comparing(cell -> cell.y)).y;
      int lowerYValue = Collections.min(cellsForBoard, Comparator.comparing(cell -> cell.y)).y;
      height =
          getDimension(lowerYValue, upperYValue, this.height);

      leftmostXValue = Collections.min(cellsForBoard, Comparator.comparing(cell -> cell.x)).x;
      int rightmostXValue = Collections.max(cellsForBoard, Comparator.comparing(cell -> cell.x)).x;
      width =
          getDimension(rightmostXValue, leftmostXValue, this.width);
    }

    private int getDimension(int lowestDimensionValue, int highestDimensionValue, int p) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) < DEFAULT_BOARD_SIZE ?
          p :
          Math.abs(highestDimensionValue - lowestDimensionValue) + 1;
    }

  }
}