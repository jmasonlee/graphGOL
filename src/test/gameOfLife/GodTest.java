package gameOfLife;

import gameOfLife.TestUtils.NeighbourGenerator;
import gameOfLife.cell.Cell;
import gameOfLife.output.BoardOutputter;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

public class GodTest {
  @Test
  public void testDoesNotReturnIsolatedCells() {
    LiveCellsGraph liveCellsGraph = new LiveCellsGraph(Arrays.asList(new Cell[]{new Cell(0,1)}));
    assertEquals(new ArrayList<>(), God.livingCellsOnNextTurn(liveCellsGraph));
  }

  @Test
  public void returnsCellsWithEnoughNeighbours(){
    List<List<Cell>> cellsAndNeighbours = new ArrayList<>();
    List<Cell> allPossibleNeighbours = NeighbourGenerator.generateAllPossibleNeighboursForCell(new Cell(0,0));
    List<Cell> cellWithTwoNeighbours = NeighbourGenerator.allCombinationsOfNeighbours(allPossibleNeighbours, 2).get(0);
    cellWithTwoNeighbours.add(new Cell(0,0));

    cellsAndNeighbours.add(cellWithTwoNeighbours);

    StringBuilder toVerify = new StringBuilder();
    cellsAndNeighbours.forEach(cellGroup -> {
      toVerify.append(BoardOutputter.createBoardOutput(cellGroup));
      toVerify.append("\n\n");

      gameOfLife.LiveCellsGraph liveCellsGraph = new gameOfLife.LiveCellsGraph(cellWithTwoNeighbours);

      toVerify.append(BoardOutputter.createBoardOutput(God.livingCellsOnNextTurn(liveCellsGraph)));
      toVerify.append("\n");
    });

    Approvals.verify(toVerify);
  }
}