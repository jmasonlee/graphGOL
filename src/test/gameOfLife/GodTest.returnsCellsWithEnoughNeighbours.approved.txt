   -1  0  1  2  3
 0|X_|X_|__|__|__|
-1|X_|__|__|__|__|
-2|__|__|__|__|__|
-3|__|__|__|__|__|
-4|__|__|__|__|__|


   -1  0  1  2  3
 0|X_|X_|__|__|__|
-1|X_|__|__|__|__|
-2|__|__|__|__|__|
-3|__|__|__|__|__|
-4|__|__|__|__|__|

